### PR TITLE
Rewrite gen_html.py.

### PR DESCRIPTION
**Pull request by spxtr**
ss

### CODE REVIEW

**Comment by rmmh** [comment]
Obvious TODOs:
- pages with no tests look strange (kubelet-gce-e2e-ci)
- each table needs a header with an explanation
- there's no indication of time
- we'll need to dockerize the job or install jinja2 on all the workers

Things we should think about:
- tooltips for every column type? What's the difference between an "unstable" and a "broken" test?
- should we use color to make it easier to read? 
- Are prefixes useful or confusing?

**Comment by rmmh** [comment]
Very good cleanup!

You should use the SVG for the logo: https://github.com/kubernetes/kubernetes/blob/master/logo/logo_with_border.svg

File: jenkins/test-history/gen_html.py [hunk 1]
```diff
@@ -14,340 +14,218 @@
...
+    'latest_failure',
+    'tests',
+    'stable',
+    'unstable',
+    'broken',
+])
 
 
-def gubernator_url(suite, build_number, test_name):
+def gubernator_url(bucket, job, build):
```
Comment (rmmh) [inline jenkins/test-history/gen_html.py]: can you retain the test_name link? on builds with many different failures, it saves time.

File: jenkins/test-history/gen_html.py [hunk 2]
```diff
@@ -14,340 +14,218 @@
...
+        stable,
+        unstable,
+        broken)
+    tests = sorted(tests, key=lambda t: (t['failed'], t['passed']), reverse=True)
+    return job_summary, tests
+
+
+def list_jobs(data):
+    """Generates (bucket, job, job_data) tuples."""
+    for bucket in data['buckets']:
```
Comment (rmmh) [inline jenkins/test-history/gen_html.py]: `for bucket in data['buckets'].itervalues()`

File: jenkins/test-history/gen_html.py (hunk 1 above)
Comment (spxtr) [inline jenkins/test-history/gen_html.py]: Yeah, might as well.

**Comment by spxtr** [comment]
> we'll need to dockerize the job or install jinja2 on all the workers

How about dockerize and also 2to3 :)

**Comment by spxtr** [comment]
> pages with no tests look strange (kubelet-gce-e2e-ci)

I no longer generate these.

> each table needs a header with an explanation

Done.

> there's no indication of time

It says "Last Updated Thu May 12 13:15:44 2016" at the top of index.html.

> we'll need to dockerize the job or install jinja2 on all the workers

Dockerized. Still on python2 though :(

> You should use the SVG for the logo

Done.

> can you retain the test_name link? on builds with many different failures, it saves time.

Done.

> `for bucket, jobs in data['buckets'].iteritems(): ...`

Done.

File: jenkins/test-history/gen_history [hunk 3]
```diff
@@ -27,13 +27,12 @@ readonly jsonpath="gs://${bucket}/logs/$(date +%F).json"
...
-# Create JSON report
-time python gen_json.py \
-  --jobs_dirs gs://kubernetes-jenkins/logs/ gs://kube_azure_log/ gs://rktnetes-jenkins/logs/ \
-  "--match=^kubernetes|kubernetes-build|kubelet-gce-e2e-ci"
-
-# Create static HTML reports out of the JSON
-python gen_html.py --output-dir=static --input=tests.json
+docker run --rm -v "$(pwd):/test-history" -w="/test-history" python:2.7 \
+    pip install -r requirements.txt && \
+    time python gen_json.py \
```
Comment (rmmh) [inline jenkins/test-history/gen_history]: this isn't executing inside the docker container. you want something like:

    docker run -i --rm -v "$(pwd):/test-history" -w="/test-history" python:2.7 bash <<EOF
    pip install -r requirements.txt && \
    time python gen_json.py ...
    EOF

**Comment by rmmh** [comment]
I mean you can't see the test execution time, where you could before. It's helpful for seeing the impact of flaky jobs.

**Comment by spxtr** [comment]
Added average duration.

File: jenkins/test-history/gen_history (hunk 3 above)
Comment (spxtr) [inline jenkins/test-history/gen_history]: Whoops, should be fixed now.

**Comment by spxtr** [comment]
@travisnewhouse this PR collides with yours.

**Comment by spxtr** [comment]
@rmmh PTAL

File: jenkins/test-history/gen_history [hunk 4]
```diff
@@ -28,16 +27,16 @@ readonly jsonpath="gs://${bucket}/logs/$(date +%F).json"
...
-time python gen_json.py \
-  --buckets=buckets.json \
-  "--match=^kubernetes|kubernetes-build|kubelet-gce-e2e-ci"
-
-# Create static HTML reports out of the JSON
-python gen_html.py \
-    --output-dir=static \
-    --input=tests.json \
-    --buckets=buckets.json
+docker run --rm -v "$(pwd):/test-history" -w="/test-history" python:2.7 bash -c "\
```
Comment (rmmh) [inline jenkins/test-history/gen_history]: pass the `-i` flag so any errors are visible instead of silently buried in a docker log

Comment (spxtr) [inline jenkins/test-history/gen_history]: Done.

File: jenkins/test-history/gen_html_test.py [hunk 5]
```diff
@@ -136,43 +174,15 @@ def check(args):
...
-            gen_html.main(tests_json, buckets_json, temp_dir)
-            for page in (
-                    'index',
-                    'tests-kubernetes',
-                    'suite-kubernetes-release',
-                    'suite-kubernetes-debug'):
-                self.assertTrue(os.path.exists('%s/%s.html' % (temp_dir, page)))
-        finally:
-            shutil.rmtree(temp_dir)
+        pass
```
Comment (rmmh) [inline jenkins/test-history/gen_html_test.py]: please fill this in. the previous test should still be applicable.

Comment (spxtr) [inline jenkins/test-history/gen_html_test.py]: whoops! I kept the in-between.

Comment (spxtr) [inline jenkins/test-history/gen_html_test.py]: Done.

File: jenkins/test-history/gen_html_test.py [hunk 6]
```diff
@@ -27,83 +28,121 @@
...
+        'gs://rktnetes-jenkins/logs/': {
+            'kubernetes-release': {
+                '7': {'tests': [{'name': 0, 'time': 0.0, 'skipped': True}]},
+            },
+        },
+        'gs://kube_azure_log/': {},
+    },
 }
 
+TEST_BUCKETS_DATA = """{
```
Comment (rmmh) [inline jenkins/test-history/gen_html_test.py]: why are you making this a string?

Comment (spxtr) [inline jenkins/test-history/gen_html_test.py]: I suppose it didn't really need to be, just need to do a `json.dumps` where the string is expected. Fixed.

**Comment by rmmh** [comment]
LGTM

**Comment by lavalamp** [comment]
Where can I view this in action?

**Comment by lavalamp** [comment]
found it, nvm :) (http://storage.googleapis.com/kubernetes-test-history/static/index.html)

**Pull request merged**